From f982d2fbf0c5c6dca3b62d3bd675645915f76a7a Mon Sep 17 00:00:00 2001
From: David Milburn <dmilburn@redhat.com>
Date: Mon, 5 May 2014 21:04:02 -0400
Subject: [block] nvme: CPU hot plug notification

Message-id: <1399323848-28227-7-git-send-email-dmilburn@redhat.com>
Patchwork-id: 79393
O-Subject: [RHEL7.1 PATCH BZ 1081734 06/12] NVMe: CPU hot plug notification
Bugzilla: 1081734
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

From: Keith Busch <keith.busch@intel.com>

Registers with hot cpu notification to rebalance, and potentially allocate
additional, io queues.

Signed-off-by: Keith Busch <keith.busch@intel.com>
Signed-off-by: Matthew Wilcox <matthew.r.wilcox@intel.com>
(cherry picked from commit 33b1e95c90447ea73e37e837ea0268a894919f19)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/block/nvme-core.c b/drivers/block/nvme-core.c
index 0bda993..e5ccffe 100644
--- a/drivers/block/nvme-core.c
+++ b/drivers/block/nvme-core.c
@@ -2084,6 +2084,19 @@ static size_t db_bar_size(struct nvme_dev *dev, unsigned nr_io_queues)
  return 4096 + ((nr_io_queues + 1) * 8 * dev->db_stride);
 }
 
+static int nvme_cpu_notify(struct notifier_block *self,
+    unsigned long action, void *hcpu)
+{
+ struct nvme_dev *dev = container_of(self, struct nvme_dev, nb);
+ switch (action) {
+ case CPU_ONLINE:
+ case CPU_DEAD:
+  nvme_assign_io_queues(dev);
+  break;
+ }
+ return NOTIFY_OK;
+}
+
 static int nvme_setup_io_queues(struct nvme_dev *dev)
 {
  struct nvme_queue *adminq = raw_nvmeq(dev, 0);
@@ -2162,6 +2175,11 @@ static int nvme_setup_io_queues(struct nvme_dev *dev)
  nvme_free_queues(dev, nr_io_queues + 1);
  nvme_assign_io_queues(dev);
 
+ dev->nb.notifier_call = &nvme_cpu_notify;
+ result = register_hotcpu_notifier(&dev->nb);
+ if (result)
+  goto free_queues;
+
  return 0;
 
  free_queues:
@@ -2439,6 +2457,7 @@ static void nvme_dev_shutdown(struct nvme_dev *dev)
  int i;
 
  dev->initialized = 0;
+ unregister_hotcpu_notifier(&dev->nb);
 
  spin_lock(&dev_list_lock);
  list_del_init(&dev->node);
diff --git a/include/linux/nvme.h b/include/linux/nvme.h
index f0f95c7..15d071e 100644
--- a/include/linux/nvme.h
+++ b/include/linux/nvme.h
@@ -92,6 +92,7 @@ struct nvme_dev {
  struct kref kref;
  struct miscdevice miscdev;
  struct work_struct reset_work;
+ struct notifier_block nb;
  char name[12];
  char serial[20];
  char model[40];
-- 
1.7.1